### PR TITLE
feat: transformDeep utility function

### LINKED ISSUE #367: feat: transformDeep
Description:
- When dealing with API responses we need to be able to transform between various shapes of different data types such as objects, nested objects, arrays, arrays of objects, arrays of primitive values such as strings and arrays of nested objects

Objectives:
- Add a `transformDeep` utility function to `core/data` which deeply transforms both iterables such as objects and primitive values such as strings and numbers
   - Follow the existing directory structure in `core/data` - the utility function should be placed under `transform-deep`
   - Add unit tests for the implementation

### PR DESCRIPTION
**Pull request by nmathew98**
Add comprehensive transformDeep utility to core/data for deeply transforming API responses and complex data structures.

## Features
- Objects, nested objects, arrays, and arrays of objects
- Primitive values (strings, numbers, booleans)
- Arrays of primitive values and nested objects
- Sets and Maps
- Circular reference handling
- Maximum depth control
- Transform function receives value, key, and parent context

Includes comprehensive unit tests covering all supported data types and edge cases.

Closes #367

Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Comment by nmathew98** [comment]
@claude when it comes to data types such as objects, often times we have a certain shape which follows certain transformation rules. update the implementation of `transformDeep` such that we are able to apply different transforms according to the result of a predicate. `Map`s allow for keys of any data type

File: client/src/core/data/transform-deep/transform-deep.test.ts [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+				[complexKey, "value1"],
+				[dateKey, "value2"],
+				[numberKey, 100],
+				["stringKey", "value3"]
+			]);
+
+			const result = transformDeep(data, { transforms });
+
+			expect(result).toBeInstanceOf(Map);
+			expect(result.get(complexKey)).toBe("transformed_value1");
```
Comment (nmathew98) [inline client/src/core/data/transform-deep/transform-deep.test.ts]: @claude this test is failing with the following error:

```
AssertionError: expected undefined to be 'transformed_value1' // Object.is equality - Expected: "transformed_value1" + Received: undefined ❯ src/core/data/transform-deep/transform-deep.test.ts:448:35
```

File: client/src/core/data/transform-deep/transform-deep.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+		const result = conform(value, context);
+
+		visited.set(value, result);
+
+		return result;
+	};
+
+	const result = deepTransform(value);
+
+	// TODO: can't think of a better way to handle this so that
```
Comment (nmathew98) [inline client/src/core/data/transform-deep/transform-deep.ts]: @claude fix this issue: the gist of it is that the value we are trying to transform might contain circular references.

it is tricky to solve this problem because a circular reference means that the result of something higher up the call stack relies on the result of something lower on the stack. the problem is solved at the moment by placeholder value, a `ref` which is repointed after the entire transformation is completed

the drawback is that the value of the circular reference can only be accessed via `ref` which is not ideal because the transformation is no longer predictable. the only transformations which should be applied are those by matching conformers. conformers applying some arbitrary transformation on matching values

the solution we're looking for is to be able to access the circular references directly instead of via `ref`.

for example, given an object:
```
{
   a: string;
   self: this; // circular reference
}
```

applying `transformDeep` as it stands would result in an object that looks like, note that to access the circular reference we have to do it via `.self.ref`:
```
{
   a: string;
   self: {
      // ... other fields
      ref: this; // circular reference
   }
}
```

instead, what we are looking for is the ability to access circular references like this after applying `transformDeep`:
```
{
   a: string;
   self: this; // circular reference
}
```

do not commit anything if you don't have a satisfactory answer. existing functionality should not be broken

Comment (nmathew98) [reply client/src/core/data/transform-deep/transform-deep.ts]: @claude thats great! could you commit the result if it works. also update tests in `transform-deep.test.ts` on `L365`

**Comment by nmathew98** [comment]
@claude you have not committed the result of fixing the issue. to restate the issue:

fix the `TODO` on `L129` of `transform-deep.ts`: the gist of it is that the value we are trying to transform might contain circular references.

it is tricky to solve this problem because a circular reference means that the result of something higher up the call stack relies on the result of something lower on the stack. the problem is solved at the moment by placeholder value, a `ref` which is repointed after the entire transformation is completed

the drawback is that the value of the circular reference can only be accessed via `ref` which is not ideal because the transformation is no longer predictable. the only transformations which should be applied are those by matching conformers. conformers applying some arbitrary transformation on matching values

the solution we're looking for is to be able to access the circular references directly instead of via `ref`.

for example, given an object:
```
{
   a: string;
   self: this; // circular reference
}
```

applying `transformDeep` as it stands would result in an object that looks like, note that to access the circular reference we have to do it via `.self.ref`:
```
{
   a: string;
   self: {
      // ... other fields
      ref: this; // circular reference
   }
}
```

instead, what we are looking for is the ability to access circular references like this after applying `transformDeep`:
```
{
   a: string;
   self: this; // circular reference
}
```

existing functionality should not be broken

File: client/src/core/data/transform-deep/transform-deep.ts [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+				hasCircularReference,
+			},
+			enumerable: false,
+		});
+	}
+
+	return result;
+};
+
+const isRef = (x: unknown) => Boolean((x as Record<string, any>)?.ref);
```
Comment (nmathew98) [inline client/src/core/data/transform-deep/transform-deep.ts]: @claude remove `isRef`, `isCircularReference` and `unwrap` as they are no longer being used. also update tests in `transform-deep.test.ts` to remove assertions which use these

File: client/src/core/data/transform-deep/transform-deep.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+			visited.set(value, result);
+			processing.delete(value);
+		}
+
+		return result;
+	};
+
+	const result = deepTransform(value);
+
+	// Phase 2: Replace all placeholders with direct references
```
Comment (nmathew98) [inline client/src/core/data/transform-deep/transform-deep.ts]: @claude now we have are able to access circular reference placeholders directly

improve upon this so that we do not have to traverse the entire tree again to replace placeholders

do not update unit tests in `transform-deep.test.ts`, existing functionality should not be broken
